Add KerrSchild analytic solution

markscheel opened the pull request:
## Proposed changes

Adds the Kerr-Schild solution and a test for it.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`


### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...
-->

markscheel committed:
Add KerrSchild.

markscheel committed:
Move some GH tags into gr namespace and different files.

Tags are now simple.  We can add more tmpl magic to make these into
ComputeItems when the time comes to do this.
Need a DataType template parameter to deal with Tensor<double...>.

nilsdeppe commented:
I volunteer @osheamonn and @markscheel to review this first and deal with the tags situation.

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,109 @@
+  static constexpr OptionString help{"Black hole in Kerr-Schild coordinates"};
+
+  KerrSchild(double mass, Spin::type spin, Center::type center,
+             const OptionContext& context = {}) noexcept;
```
Are optioncontext's still needed if they're not used @wthrowe ?

wthrowe replied:
No, but as mentioned in other comments this one should be used.

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,109 @@
+      noexcept;
+
+  // clang-tidy: google-runtime-references
+  void pup(PUP::er& /*p*/) noexcept {}  // NOLINT
```
Shouldn't the `pup` function pup the mass, center and spin?

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+      spin_(std::move(spin))  // NOLINT
+{
+  const double spin_magnitude = magnitude(spin_);
+  if (spin_magnitude > 1.0) {
```
Is it possible for this to determined as a PARSE_ERROR or something along those lines? I'm not sure.

markscheel replied:
I don't know.  This constructor, however, should be callable without using Options.

wthrowe replied:
Use PARSE_ERROR.  It will do something reasonable if the context is default constructed.

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+}
+
+Variables<KerrSchild::Tags> KerrSchild::solution(
+    const tnsr::I<DataVector, 3>& x, const double /*t*/) const noexcept {
```
This function has a lot of temporaries which can be made const by initialising using a lambda.

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+    a_squared += square(s);
+  }
+
+  auto x_minus_center = x;
```
Const using lambda

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+  const DataVector a_dot_x_over_rsquared = a_dot_x / r_squared;
+
+  const DataVector deriv_log_r_denom = 0.5 / (r_squared - half_xsq_minus_asq);
+  tnsr::i<DataVector, 3> deriv_log_r;
```
const using lambda

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+  const DataVector H_denom = 1.0 / (square(r_squared) + a_dot_x_squared);
+  const DataVector H = mass_ * sqrt(r_squared) * r_squared * H_denom;
+
+  tnsr::i<DataVector, 3> dH;
```
const using lambda.

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+    }
+  }
+
+  tnsr::i<DataVector, 3> a_cross_x;
```
const using lambda.

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+  const DataVector denom = 1.0 / (r_squared + a_squared);
+  const DataVector r = sqrt(r_squared);
+
+  tnsr::i<DataVector, 3> null_form;
```
const

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+    }
+  }
+
+  tnsr::ij<DataVector, 3> deriv_null_form;
```
const

osheamonn commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,187 @@
+  Index<3> extents(grid_size);
+
+  // Set up map
+  CoordinateMaps::AffineMap x_map(-1., 1., lower_bound[0], upper_bound[0]);
```
Can the map construction simplified by `make_coordinate_map`?

osheamonn commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,187 @@
+      coord_map(xyz_map);
+
+  // Set up coordinates
+  const auto x_logical = logical_coordinates(extents);
```
inline into the next line

osheamonn commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,187 @@
+  }
+
+  const DataVector r = magnitude(x);
+  auto lapse_test = make_with_value<Scalar<DataVector>>(x, 0.0);
```
I think appending `_test` doesn't make it clear whether this the expected lapse or the computed. I prefer to prepend `expected_`, I don't feel too strongly though, what do other people think?

osheamonn commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,187 @@
+  }
+  CHECK_ITERABLE_APPROX(g, g_test);
+
+  auto dg_test =
```
The casing is inconsistent with `d_g`

osheamonn commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.cpp:
```diff
@@ -0,0 +1,284 @@
+  Index<3> extents(grid_size_each_dimension);
+
+  // Set up map
+  CoordinateMaps::AffineMap x_map(-1., 1., lower_bound[0], upper_bound[0]);
```
I think this can be simplified with make_coordinate_map

osheamonn commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.cpp:
```diff
@@ -0,0 +1,284 @@
+    for (size_t b = 0; b < 4; ++b) {
+      for (size_t i = 0; i < 3; ++i) {
+        for (size_t s = 0; s < data_size; ++s) {
+          CHECK(d_psi.get(i, a, b)[s] == numerical_approx(phi.get(i, a, b)[s]));
```
Will CHECK_ITERABLE_APPROX work here?

osheamonn commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.cpp:
```diff
@@ -0,0 +1,284 @@
+  for (size_t a = 0; a < 4; ++a) {
+    for (size_t b = 0; b < 4; ++b) {
+      for (size_t s = 0; s < data_size; ++s) {
+        CHECK(dt_psi.get(a, b)[s] == numerical_approx(0.0));
```
Can probably eliminating the for loops using CHECK_ITERABLE_APPROX along with make_with_value

osheamonn commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.cpp:
```diff
@@ -0,0 +1,284 @@
+
+// Need explicit instantiation of derivatives.
+
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.cpp"
```
Why is this include needed here? If this is really the way to do this then I'd add another line after it.

wthrowe replied:
This is how it is designed to be used, but actually looking at this I think it's kind of an invitation to have everything blow up in our faces.  In particular, if the included file contains anything that isn't templated (or inline) we'll likely get linking errors.

PartialDerivative.cpp needs to follow all the rules for header files, and I think we should rename it to something ending in hpp to avoid confusion.

osheamonn commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.cpp:
```diff
@@ -0,0 +1,284 @@
+// Instantiate for KerrSchild
+
+#include "PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp"
+template void
```
newline

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+    ERROR("Spin magnitude must be < 1. Given spin: "
+          << spin_ << " with magnitude " << spin_magnitude);
+  }
+  if (mass_ < 0.0) {
```
Same here

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+  get(get<lapse>(result)) = sqrt(lapse_squared);
+
+  {
+    const DataVector fac =
```
Since this is only used in one calculation should it just be inlined to prevent the memory allocation?

markscheel replied:
Don't know.  I kept it out because if inlined it would be computed 3 times instead of once.

osheamonn replied:
Oh yeah of course, nevermind.

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+  }
+
+  {
+    const DataVector fac = -2.0 * H * null_vector_0 * lapse_squared;
```
Same here.

osheamonn commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,109 @@
+namespace EinsteinSolutions {
+
+/*!
+ * \ingroup EinsteinSolutions
```
EinsteinSolutionsGroup

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,109 @@
+    using type = double;
+    static constexpr OptionString help = {"Mass of the black hole"};
+    static type default_value() { return 1.; }
+    static type lower_bounds() { return 0.; }
```
`lower_bound` (no 's')

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,109 @@
+    static constexpr OptionString help = {"The [x,y,z] spin of the black hole"};
+    static type default_value() { return {{0., 0., 0.}}; }
+    static type lower_bounds() { return {{-1., -1., -1.}}; }
+    static type upper_bounds() { return {{1., 1., 1.}}; }
```
I think these do a lexicographic comparison.  Since you have to check the magnitude anyway in the constructor I'd just drop the bounds here.

wthrowe reviewed (changes requested):
I've only commented on the option parsing stuff.

gsb76 commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,109 @@
+ * \f$H\f$ and \f$l_\mu\f$ are known functions of the coordinates
+ * and of the mass and spin vector.
+ *
+ * The following are input file options that can be specified:
```
I think this can be moved into the help OptionString (line 53)

mar-celine commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,187 @@
+
+  // Set up map
+  CoordinateMaps::AffineMap x_map(-1., 1., lower_bound[0], upper_bound[0]);
+  CoordinateMaps::AffineMap y_map(-1., 1., lower_bound[1], upper_bound[1]);
```
Can use `using AffineMap = CoordinateMaps::AffineMap;`
 and `using AffineMap3D =
      CoordinateMaps::ProductOf3Maps<AffineMap, AffineMap, AffineMap>;`

mar-celine commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,187 @@
+  for (size_t k = 0; k < 3; ++k) {
+    for (size_t i = 0; i < 3; ++i) {
+      for (size_t j = i; j < 3; ++j) {  // Symmetry
+        CHECK_ITERABLE_APPROX(d_g.get(k, i, j), dg_test.get(k, i, j));
```
I think `CHECK_ITERABLE_APPROX` here makes the loops unnecessary

mar-celine commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,187 @@
+  // Parameters
+  //   ...for KerrSchild solution
+  const double mass = 1.01;
+  const std::array<double, 3> spin = {{0.1, 0.2, 0.3}};
```
General question: is it any better to do `spin{{0.1, 0.2, 0.3}}` instead?

nilsdeppe replied:
The code has the same meaning, but I prefer `spin{{...}}}` because that makes it clear it is a constructor call, not an assignment operator call. This seems to be (understandably) confusing to people newer to C++. That is, they think because of the `=` it must use `operator=`, when really the compiler just calls the constructor.

mar-celine commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,187 @@
+  EinsteinSolutions::KerrSchild solution(mass, spin, center);
+  verify_time_independent_einstein_solution(
+      solution, grid_size, lower_bound, upper_bound,
+      std::numeric_limits<double>::epsilon() * 100000);
```
I would go with `1.0e5`

mar-celine commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.cpp:
```diff
@@ -0,0 +1,284 @@
+  // Constraint-damping parameters: Set to arbitrary values.
+  // gamma1 = 0 and -1 are special, so don't choose that one.
+  const auto gamma0 = make_with_value<Scalar<DataVector>>(x, 1.0);
+  const auto gamma1 = make_with_value<Scalar<DataVector>>(x, 1.0);
```
`const auto& gamma1 = gamma0;`

osheamonn replied:
Why use a reference? I think this is correct as is since the RHS will be temporary.

mar-celine replied:
:+1:

wthrowe replied:
It would be an optimization, as it avoids an allocation, but since this is just a test that shouldn't matter.  It's probably clearer as is, so I'd leave it.

gsb76 commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.cpp:
```diff
@@ -0,0 +1,284 @@
+  const auto gamma2 = make_with_value<Scalar<DataVector>>(x, 1.0);
+
+  // Compute RHS of generalized harmonic Einstein equations.
+  auto dt_psi =
```
`const auto zero_aa_tensor = make_with_value<tnsr::aa<DataVector, 3, Frame::Inertial>>(x, 0.0);`
`const auto& dt_psi = zero_aa_tensor;`
`const auto& dt_pi = zero_aa_tensor;`

mar-celine commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,172 @@
+  }
+
+  const constexpr double null_vector_0 = -1.0;
+  const DataVector lapse_squared = 1.0 / (1. + 2. * H * square(null_vector_0));
```
`1.0 / (1.0 + 2.0`

osheamonn replied:
Our code review guidelines explicitly say that `1.` or `1.0` is fine.

mar-celine replied:
I think I remember a pull request where we didn't want to mix and match these in the same line, but I might be wrong.

wthrowe replied:
I agree, let's try to be locally consistent.  (I don't care which way you change it.)

mar-celine commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,187 @@
+    }
+  }
+
+  const DataVector r = magnitude(x);
```
Would it help to precompute `cube(r)` and `square(r)` (or their reciprocals) here? Since `cube(r)` resolves to `r * r * r` and that is done a lot in the following tests.

mar-celine commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,187 @@
+    "Unit.PointwiseFunctions.AnalyticSolutions.EinsteinSolution.KerrSchildSpin",
+    "[PointwiseFunctions][Unit]") {
+  ERROR_TEST();
+  EinsteinSolutions::KerrSchild solution(1.0, {{1.0, 1.0, 1.0}},
```
Should there also be tests for constructing via Options?

markscheel replied:
Yes, there are now.

markscheel commented:
I've addressed all the issues except I don't know how to CHECK_ITERABLE_APPROX for a different epsilon than the usual one.

KerrSchild now works with doubles as well as DataVectors.

I'm getting strange compiler warnings (gcc only) with PARSE_ERROR.

And I have an additional test I didn't commit: constructing KerrSchild with options with spin magnitude > 1 doesn't appear to produce any error message, it just throws SIGABORT.

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,244 @@
+      l_a_squared += square(s);
+    }
+    return l_a_squared;
+  }(spin_);
```
How about `a_squared = std::inner_product(spin.begin(), spin.end(), spin.begin(), 0.)`

I don't know whether a similar can be used later when there are Tensors and stuff in use, but here we just have arrays of doubles.

wthrowe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,122 @@
+    static type default_value() { return {{0., 0., 0.}}; }
+  };
+  using options = tmpl::list<Mass, Spin, Center>;
+  using type = KerrSchild;  // This class is its own options creator.
```
What is this for?  I don't think option parsing uses it.

markscheel replied:
In Test_KerrSchild.cpp, `opts.get<EinsteinSolutions::KerrSchild>()` fails to compile if I leave this out.

In the example I see in the code, a class creatable by options uses a helper class that contains a `type` type alias, and then one uses the helper class as the template parameter to opts.get.  But here it seems easier to me to just have KerrSchild itself be the helper class, so you avoid creating an extra class.  (The example I am talking about is in Test_CustomTypeConstruction.cpp, where there is a class `CreateFromOptions` and then a helper class `CFO` that contains `using type = CreateFromOptions<int>`.  Here KerrSchild is the equivalent of `CreateFromOptions`, and instead of making a new class equivalent to `CFO` and then worrying about what I should name it, I just put the type alias into `KerrSchild`. The example is slightly different in that `CreateFromOptions` is a template class, so in that case `CFO` must be a separate class because it can't be a template.)  Is there an easier way to do this, avoiding the type alias in `KerrSchild`?

wthrowe replied:
It's an interesting thought, and I've been going back and forth on whether I like this idea.  It has the definite advantage of requiring less code, and that could be cleaned up even more by defaulting the `type` parameter to the option struct class.  (That couldn't create weird types because plain option structs are not constructible in the option parsing sense.)

I've decided I'm going to discourage it, though.  The two classes have different purposes and different restrictions.  The class `help` should give a reasonably detailed description of the class (which in this case happens to be short), but the option struct `help` has severe length limits (56 characters) and I worry this will encourage people to try to fit their class `help` into those limits.  More importantly, the messages have different purposes: the class `help` says what the class *is*, but the option `help` says what it is *for*.  There may be several option structs that create the same class.  An entry in an input file described as "KerrSchild: Black hole in Kerr-Schild coordinates" gives no hint as to what it is for and probably should read "BackgroundKerrSchild: The background metric" or "InitialData: Initial data parameters" or similar.

(Digression: I'm uncertain about whether we should include the class in the option name.  The class is clear from the help text, but not from just reading the input file.)

For testing purposes I've been meaning to write something like `test_factory_creation` for non-factory things.  I'll see if I can get that in today.

markscheel replied:
Ok, I am now using a separate option parsing struct.  I hadn't thought about multiple option parsing structs creating the same class.

wthrowe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,299 @@
+      make_with_value<tnsr::I<DataType, 3, Frame::Inertial>>(x, 0.0);
+  for (size_t i = 0; i < 3; ++i) {
+    expected_shift.get(i) =
+        2.0 * mass * x.get(i) * one_over_r_squared * square(get<>(lapse));
```
No `<>`, and again in a few lines.

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,122 @@
+ * \brief Kerr black hole in Kerr-Schild coordinates
+ *
+ * \details
+ * The metric is \f$g_{\mu,\nu} = \eta_{\mu\nu} + 2 H l_\mu l_\nu\f$,
```
remove comma in g_{\mu\nu}

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,122 @@
+ * where \f$\eta_{\mu\nu}\f$ is the Minkowski metric, \f$H\f$ is a scalar
+ * function, and \f$l_\mu\f$ is the outgoing null vector.
+ * \f$H\f$ and \f$l_\mu\f$ are known functions of the coordinates
+ * and of the mass and spin vector.
```
Could you try run the PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchildCoords.tex file from SpEC through the python code ($SPECTRE_HOME/tools/latex2doc.py) that tries to transcribe it into something Doxygen can handle and then add the result? That document is quite useful and it would be nice to retain that info in SpECTRE.

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,122 @@
+ * and of the mass and spin vector.
+ *
+ * The following are input file options that can be specified:
+ * Mass (default: 1.), Center (default: {0,0,0}), Spin (default: {0,0,0})
```
this would probably look better as a bullet list

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,122 @@
+    static type default_value() { return 1.; }
+    static type lower_bound() { return 0.; }
+  };
+  struct Spin {
```
clarify if this is the dimensionless or dimensionful spin

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp:
```diff
@@ -0,0 +1,122 @@
+#include <array>
+
+#include "DataStructures/Tensor/TypeAliases.hpp"
+#include "DataStructures/Variables.hpp"
```
remove

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/Magnitude.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
```
remove

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+#include "PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp"
+
+#include <numeric>
+#include <pup.h>
```
include utility

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+    const tnsr::I<DataType, 3>& x, const double /*t*/) const noexcept {
+  const auto a_squared =
+      std::inner_product(spin_.begin(), spin_.end(), spin_.begin(), 0.);
+  //-:  const auto a_squared = [](const auto& spin) {
```
remove commented code

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+  //-:    return l_a_squared;
+  //-:  }(spin_);
+
+  const auto x_minus_center = [](const auto& coord, const auto& center) {
```
replace auto in lambda with actual types. The lambda does not need to be a template function

mark noexcept

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+
+  const DataType deriv_log_r_denom = 0.5 / (r_squared - half_xsq_minus_asq);
+
+  const auto deriv_log_r = [&]() {
```
explicitly specify what you are capturing, mark noexcept

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+  const DataType H_denom = 1.0 / (square(r_squared) + a_dot_x_squared);
+  const DataType H = mass_ * sqrt(r_squared) * r_squared * H_denom;
+
+  const auto deriv_H = [&]() {
```
capture what you need, mark noexcept

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+
+  const auto deriv_H = [&]() {
+    auto l_deriv_H = make_with_value<tnsr::i<DataType, 3>>(x, 0.0);
+    const DataType fac1 = H * (3.0 - 4.0 * square(r_squared) * H_denom);
```
fac -> `temp`, or `coef`

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+    return l_deriv_H;
+  }();
+
+  const auto a_cross_x = [](const auto& a, const auto& coord) {
```
noexcept, and specify the types, no need for a generic lambda

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+  const DataType denom = 1.0 / (r_squared + a_squared);
+  const DataType r = sqrt(r_squared);
+
+  const auto null_form = [&]() {
```
capture what you need, mark noexcept

Just do that everywhere

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+
+  const auto null_form = [&]() {
+    auto l_null_form = make_with_value<tnsr::i<DataType, 3>>(x, 0.0);
+    const DataType fac = a_dot_x / r;
```
`temp` or `coef`

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+  }();
+
+  const constexpr double null_vector_0 = -1.0;
+  const DataType lapse_squared = 1.0 / (1.0 + 2.0 * H * square(null_vector_0));
```
We're really confused by `null_vector_0`'s use on this line (`square`) If this changes for boosted solutions a comment would be useful

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+  const DataType lapse_squared = 1.0 / (1.0 + 2.0 * H * square(null_vector_0));
+
+  // Eventually will replace this by defining
+  // MakeWithValueImpl<TaggedTuple<...>,Tensor<...>>.
```
either do this or file an issue and remove the comment

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+      sqrt(lapse_squared);
+
+  {
+    const DataType fac =
```
awe fac...

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,248 @@
+  }
+
+  {
+    const DataType fac = -2.0 * H * null_vector_0 * lapse_squared;
```
just rename the fac's everywhere

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.hpp:
```diff
@@ -0,0 +1,18 @@
+
+#include <array>
+
+/// Determines if the given `solution` is a time-independent solution
```
add to TestingFrameworkGroup

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.hpp:
```diff
@@ -0,0 +1,18 @@
+
+#pragma once
+
+#include <array>
```
include cstddef

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.cpp:
```diff
@@ -0,0 +1,261 @@
+#include "PointwiseFunctions/GeneralRelativity/ComputeGhQuantities.hpp"
+#include "PointwiseFunctions/GeneralRelativity/ComputeSpacetimeQuantities.hpp"
+#include "PointwiseFunctions/GeneralRelativity/IndexManipulation.hpp"
+#include "Utilities/TypeTraits.hpp"
```
TMPL

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.cpp:
```diff
@@ -0,0 +1,261 @@
+}  // namespace
+
+template <typename Solution>
+void verify_time_independent_einstein_solution(
```
move this function to header file so it's simpler to use. We're not so concerned about compilation time in this case because you can compile all the cpp files for the tests in parallel anyway.

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.cpp:
```diff
@@ -0,0 +1,261 @@
+
+// Instantiate for KerrSchild
+
+#include "PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.hpp"
```
Remove this instantiation

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/VerifyEinsteinSolution.cpp:
```diff
@@ -0,0 +1,261 @@
+
+// Need explicit instantiation of derivatives.
+
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.cpp"
```
keep this instantiation

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,299 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "DataStructures/DataVector.hpp"
```
Tensor/TypeAliases.hpp and Tensor/Tensor.hpp

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,299 @@
+  // Schwarzschild solution is:
+  // H        = M/r
+  // l_mu     = (1,x/r,y/r,z/r)
+  // lapse    = (1+2M/r){-1/2}
```
there is something missing between the `)` and `{`, maybe a `^` or a `/`? and next line too

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,299 @@
+  // g_{ij}   = delta_{ij} + 2 M x_i x_j/r^3
+  // d_i H    = -Mx_i/r^3
+  // d_i l_j  = delta_{ij}/r - x^i x^j/r^3
+  // d_k g_ij = -2M x_i x_j x_k/r^5 + 4(M/r)x_(i/r [delta_{kj)}/r-x_k x_j)/r^3]
```
delete this line and just have the next line (what it's equal to)

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,299 @@
+  //   ...for grid
+  const size_t grid_size = 8;
+  const std::array<double, 3> lower_bound{{1.99, -0.01, -0.01}};
+  const std::array<double, 3> upper_bound{{2.01, 0.01, 0.01}};
```
move these further away from the coordinate axes

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,299 @@
+  //   ...for KerrSchild solution
+  const double mass = 1.01;
+  const std::array<double, 3> spin{{0.1, 0.2, 0.3}};
+  const std::array<double, 3> center{{0.01, 0.02, 0.03}};
```
move away from origin

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,299 @@
+void test_einstein_solution() {
+  // Parameters
+  //   ...for KerrSchild solution
+  const double mass = 1.01;
```
increase mass to ~1.7

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,299 @@
+
+void test_double_vs_datavector() noexcept {
+  // Parameters for KerrSchild solution
+  const double mass = 1.01;
```
again

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,299 @@
+    "[PointwiseFunctions][Unit]") {
+  test_schwarzschild<DataVector>(DataVector{1.0, 2.0, 3.0});
+  test_schwarzschild<double>(1.0);
+  test_double_vs_datavector();
```
swap order to be the same as defined above

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,299 @@
+SPECTRE_TEST_CASE(
+    "Unit.PointwiseFunctions.AnalyticSolutions.EinsteinSolution.KerrSchild",
+    "[PointwiseFunctions][Unit]") {
+  test_schwarzschild<DataVector>(DataVector{1.0, 2.0, 3.0});
```
since the DataVector and double are just used for size, let's set them to 0. or `nan` to make it obvious they're not used for anything

nilsdeppe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_KerrSchild.cpp:
```diff
@@ -0,0 +1,299 @@
+      std::numeric_limits<double>::epsilon() * 1.e5);
+}
+
+void test_double_vs_datavector() noexcept {
```
be consistent about `noexcept` or not `noexcept` in this file

markscheel committed:
Add MakeWithValueImpl<TaggedTuple<...>,T>

nilsdeppe commented on src/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/KerrSchild.cpp:
```diff
@@ -0,0 +1,237 @@
+    }
+    return l_x_minus_center;
+  }
+  ();
```
This is an interesting formatting choice. It would probably be clearer to have the parentheses on the previous line where the lambda ends. This appears to be consistent below. Is this a clang-format thing? Could you try clangformat 4 or 5 to see if it's less ridiculous?

markscheel replied:
clang-format 4 does the same thing.

nilsdeppe replied:
Okay, well such is life. I'm fine with you leaving it as is then.

nilsdeppe commented on src/DataStructures/MakeWithValue.hpp:
```diff
@@ -7,6 +7,7 @@
 #pragma once
```
Not a request at all, but want to bring it up for discussion.

Since TaggedTuple is in Utilities and we normally have the MakeWithValue impls inside the header of the data structure it makes, we run into an odd dependency. I don't think TaggedTuple should be moved to DataStructures, since it's very general and low level, though it is technically a data structure. However, I have nothing against moving MakeWithValue to Utilities, and then having the TaggedTuple implementation appended to it. I just don't think anything in Utilities should depend on things in DataStructures (if possible). Another possibility is to leave the impl for TaggedTuple and possibly other Utilities data structures in MakeWithValue.hpp, or to add another header that holds all the impls.

nilsdeppe replied:
Discussion with @kidder and @markscheel with agreement to make an issue and move MakeWithValue.hpp to Utilities later

nilsdeppe reviewed (changes requested):
I think the final commit order should probably be:

- GH tags
- make_with_value for tuple
- Add Kerr-Schild

I'm fine with a squash at this point.

nilsdeppe commented:
I've pushed the documentation to [here](http://nilsdeppe.com/spectre/pr_455) for people to look at.

wthrowe commented on tests/Unit/PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/CMakeLists.txt:
```diff
@@ -1,8 +1,13 @@
 set(SPECTRE_UNIT_EINSTEIN_SOLUTION
-  PointwiseFunctions/AnalyticSolutions/EinsteinSolutions/Test_Minkowski.cpp
+  ${TEST_DIR}/Test_Minkowski.cpp
+  ${TEST_DIR}/Test_KerrSchild.cpp
```
Alphabetize

kidder commented:
One of the builds fails the test; you may need to adjust the tolerance

kidder commented:
@wthrowe @nilsdeppe can you have a final look at this PR

wthrowe commented on src/DataStructures/MakeWithValue.hpp:
```diff
@@ -28,7 +29,7 @@ struct MakeWithValueImpl {
 /// \see MakeWithValueImpls
 template <typename R, typename T>
-SPECTRE_ALWAYS_INLINE R make_with_value(const T& input, const double value) {
+SPECTRE_ALWAYS_INLINE R make_with_value(const T& input, double value) {
```
This change looks odd, but it's minor so I won't block merging unless someone else cares more.

kidder replied:
Mark removed the const because of this clang-tidy warning
(see https://travis-ci.org/sxs-collaboration/spectre/jobs/319868864)
```
/work/spectre/src/DataStructures/MakeWithValue.hpp:32:57: warning: parameter 'value' is const-qualified in the function declaration; const-qualification of parameters only has an effect in function definitions [readability-avoid-const-params-in-decls]
SPECTRE_ALWAYS_INLINE R make_with_value(const T& input, const double value) {
                                                                                                          ^
```

kidder replied:
I agree, it seems odd
